src/lsm/grid.zig

```diff
@@ -484,10 +537,27 @@ pub fn GridType(comptime Storage: type) type {
                 }
             }
 
+            // Start a queued write if possible *before* calling the completed
+            // write's callback. This ensures that if the callback calls
+            // Grid.write_block() it doesn't preempt the queue.
+            //
+            // (Don't pop from the write queue until after the read-repairs are resolved.
+            // Otherwise their resolution might complete grid cancellation, but the replica has
+            // not released its own write iop (via callback).)
+            if (grid.write_queue.pop()) |queued_write| {
+                grid.write_block_with(iop, queued_write);
+            } else {
+                grid.write_iops.release(iop);
```

Comment:
Is there a way to restructure this code to avoid the need even for the careful order?

Reply:
Not that I can think of at the moment. (Btw, this code is not new to this PR, just moved around.)